### PR TITLE
Allow comparison of elements that live in equal fields

### PR DESCRIPTION
**Pull request by saraedum**
but not in identical C++ wrappers

### CODE REVIEW

**Comment by saraedum** [comment]
Here is the proper replacement for #45.

**Comment by videlec** [comment]
You don't want to encourage users to create equal but non identical fields... What is the use case you have in mind?

**Comment by saraedum** [comment]
In flatsurf I convert the number fields there to `renf_class` that's why there can be multiple copies of the same `renf_class`. I don't fully understand what you mean by encourage here. I would like to be able to write (this currently works):
```
renf_class K(…);
renf_class L = K; // maybe not like that but this happens when you put renf_class into a vector e.g.
assert(L == K);
```
This means that you also get:
```
renf_elem_class x(K), y(L);
assert(x.parent() == y.parent());
```
However, `x == y` throws an error at the moment telling you that `x` and `y` live in different number fields.

**Comment by videlec** [comment]
What I wanted to say with "encourage" is that I don't see any reason why you should have equal but non-identical parents. If the user is enforced to avoid creating parents, that is just better.

In your example, I don't understand why you would need `renf_class L = K;` in the first place.

Finally, checking identity versus equality makes a difference in terms of timing.

**Comment by saraedum** [comment]
> What I wanted to say with "encourage" is that I don't see any reason why you should have equal but non-identical parents. If the user is enforced to avoid creating parents, that is just better.

I see. But then we should not provide copy constructors and assignment for `renf_class` so it's much harder to shoot yourself in the foot. Then however, you cannot store `renf_class` in STL containers which might be annoying.

> In your example, I don't understand why you would need renf_class L = K; in the first place.

I return a field as the result of a function (among other stuff.) I need assignment to be able to return (could be done differently of course.) The `renf_class K` is a temporary in that function so I cannot return a reference to it. I could of course use smart pointers to return it but since `renf_class` provides the copy constructor it seems natural to me that I also get `K==L` (which is already the case at the moment.)

> Finally, checking identity versus equality makes a difference in terms of timing.

Absolutely. So, `renf_class`'s `operator==` should probably add a `if (this == &other) return true` in the beginning  to shortcut somewhat.

The above are not very strong arguments for my case I think apart from the fact that it's imho best to have the most convenient interface, i.e., all conceivable operators just work, even though it might be somewhat slower. Here is maybe a stronger reason why we might have non-unique parents eventually. Say we want to support arithmetic between elements living in different number fields. Then we need to promote elements to bigger fields that did not exist initially. We could do some caching of embeddings here but I guess it would be tricky to make the following work (without the proposed change here.)
```
renf_elem_class a,b,c; // in fields K,L,M
assert((a + b) + c) == (a + (b + c)));
```

**Comment by saraedum** [comment]
That said, if you feel strongly about this, I can work around of course.

**Comment by videlec** [comment]
I am fine merging it, but you should not create equal number fields.

File: renfxx/renf_elem_class.cpp [hunk 1]
```diff
@@ -523,7 +523,7 @@ bool renf_elem_class::operator<(const renf_elem_class & other) const noexcept
 {
     if (nf != nullptr)
     {
-        if (nf == other.nf)
+        if (other.nf != nullptr && *nf == *other.nf)
```
Comment (videlec) [inline renfxx/renf_elem_class.cpp]: Please change to first test pointer equality.

File: renfxx/renf_elem_class.cpp [hunk 2]
```diff
@@ -503,7 +503,7 @@ bool renf_elem_class::operator==(const renf_elem_class & other) const noexcept
...
```
Comment (videlec) [inline renfxx/renf_elem_class.cpp]: Please change to first test pointer equality.

File: renfxx/renf_class.cpp [hunk 3]
```diff
@@ -115,7 +115,9 @@ renf_elem_class renf_class::gen() const noexcept
 
 bool renf_class::operator==(const renf_class & other) const noexcept
 {
-    return renf_equal(this->nf, other.nf) && this->name == other.name;
+    if (this == &other) return true;
+    return (this->nf == other.nf || renf_equal(this->nf, other.nf))
```
Comment (videlec) [inline renfxx/renf_class.cpp]: Checking `this->nf == other->nf` would have been sufficient. The distinction between the C++ class and the pointed `nf` attribute should not be relevant here. I would just remove line 118 `if (this == &other) return true;`

**Pull request merged**